packages/liveblocks-core/src/lib/create-store.ts

```diff
@@ -25,18 +28,50 @@ export function createStore<T>(initialState: T): Store<T> {
    * Update the current state and notify all the subscribers of the update.
    */
   function set(callback: (currentState: T) => T) {
-    const newState = callback(state);
-    if (state === newState) {
-      return;
+    const oldState = state;
+    const newState = callback(oldState);
+
+    if (newState !== oldState) {
+      state = newState;
+      dirty = true;
     }
 
-    state = newState;
+    if (notifyImmediately) {
+      notify();
+    }
+  }
+
+  function notify() {
+    if (!dirty) {
+      return;
+    }
 
+    dirty = false;
     for (const subscriber of subscribers) {
       subscriber(state);
     }
   }
 
+  /**
+   * While the callback is running, do not notify any subscribers. Keep
+   * collecting (potentially multiple) .set() updates to the store, and only
+   * notify subscribers at the end of the callback.
+   */
+  function batch(cb: () => void): void {
+    if (notifyImmediately === false) {
+      // Already in a batch, make this inner batch a no-op
+      return cb();
+    }
```

Comment:
Is this to guard against nested batch calls and unwanted asynchronous operations in the callback?

Reply:
Yep! If you'd ever do:

```tsx
store.batch(() => {
  store.set();
  store.batch(() => {
    store.set();
    store.set();
  });
  store.set();
});
```

Then only the outermost `.batch()` will trigger one notify/rerender.

Async functions should never be used here. This only supports synchronous updates.

Reply:
You can think of "batching" as "wait to notify subscribers until I'm done with all these state changes first". The benefit is that you can turn any statements inside the `.batch()` into a helper function to mutate state, without having to worry about accidentally triggering subscription notifications. The behavior is very much like how our public [`Room.batch`](https://liveblocks.io/docs/api-reference/liveblocks-client#Room.batch) API works.